ethcore/src/client/client.rs

```diff
@@ -1438,15 +1443,28 @@ impl RegistryInfo for Client {
 
 impl CallContract for Client {
 	fn call_contract(&self, block_id: BlockId, address: Address, data: Bytes) -> Result<Bytes, String> {
+		self.call_contract_from_sender(block_id, address, Address::default(), data)
+	}
+
+	fn call_contract_from_sender(&self, block_id: BlockId, address: Address, sender: Address, data: Bytes) -> Result<Bytes, String> {
 		let state_pruned = || CallError::StatePruned.to_string();
 		let state = &mut self.state_at(block_id).ok_or_else(&state_pruned)?;
 		let header = self.block_header_decoded(block_id).ok_or_else(&state_pruned)?;
 
-		let transaction = self.contract_call_tx(block_id, address, data);
+		let transaction = self.contract_call_tx_from_sender(block_id, address, sender, data);
 
-		self.call(&transaction, Default::default(), state, &header)
-			.map_err(|e| format!("{:?}", e))
-			.map(|executed| executed.output)
+		let executed = self.call(&transaction, Default::default(), state, &header).map_err(|e| format!("{:?}", e))?;
+		if let Some(exception) = executed.exception {
+			let mut error = format!("{:?}", exception);
+			if exception == VmError::Reverted {
+				let output = executed.output.clone();
+				let msg = self.try_decode_solidity_revert_msg(&output).unwrap_or_else(|| format!("{}", Token::Bytes(output)));
+				error = format!("{} {}", error, msg);
+			}
+			Err(error)
```

Comment:
To make this a little easier to follow I suggest:

```
if let Some(exception) = executed.exception {
  match exception {
    VmError::Reverted => {
      …
      Err(format!("{:?} {}", exception, msg))
  },
  _ => Err(exception.to_string()) // not sure if `From<String>` is impl'd for `VmError`
}
```

Reply:
A more general question I have is: why do this error munging here rather than in the `Debug` impl for `VmError`? Couldn't we add the decoding attempt there and essentially keep this code to a simple `map_err(|e| format!("{:?}", e))` like it was before? Seems like this is spreading out the logic into several places and I'm not sure I understand why it's worth it.

Reply:
Unfortunately there is no access from `VmError` to tx `output` which we need for decoding error. Alternatively we could put all this error munging under `Executed` in some function like `format_error_and_decode_ouput`. What do you think about it @dvdplm ?

Reply:
> Unfortunately there is no access from VmError to tx output 

Can we change that maybe? Like, `VmError::Reverted(tx_output)`. If not, then I'm not opposed to having a `format_error_and_decode_ouput` helper like you suggest.